### PR TITLE
Merge com.ibm.management into jdk.management

### PR DESCRIPTION
**Pull request by mikezhang1234567890**
The com.ibm.management module is merged into jdk.management so users
that require jdk.management also have access to OpenJ9 specific MXBeans.

Changes are:
1. Move all source files from com.ibm.management module to under
jdk.management
2. In the module-info.java.extra files of the modules java.base,
java.management, and jdk.management, correct exports statements to use
the moved packages
3. Add entries to j9module.xml to fix modules that export to
old module name
4. Add jdk.jconsole and openj9.sfj modules to be processed by
j9modules.xml, since these modules reference old com.ibm.management name
5. Add a step in runtime/jvm.xml.ftl to delete com.ibm.management jmod
file in order to work with composing step when building JVM using old
method
6. Changed test material to use jdk.management module
7. Changed removeexports macro in j9modules.xml to removeexportsoropens
to handle both lines in module-info.java files
8. Added replaceexportsoropens macro in j9modules.xml to handle exports
or opens lines based on module name

Also includes change in https://github.com/ibmruntimes/openj9-openjdk-jdk9/pull/77

Signed-off-by: Mike Zhang <mike.h.zhang@ibm.com>

### CODE REVIEW

**Comment by mikezhang1234567890** [comment]
Reviewers: @JasonFengJ9 @DanHeidinga

**Comment by DanHeidinga** [comment]
@mikezhang1234567890 Can the openj9.management classes be merged into jdk.management?  That way users that `require jdk.management` will have access to the openj9 mxbeans as well, easing the work to support both JVMs and reducing the number of patches we need.

**Comment by JasonFengJ9** [comment]
In reply to @DanHeidinga comment above, technically module `openj9.management` classes can be merged into `jdk.management`, actually J9 already adds some classes within `/jcl/src/jdk.management/share/classes` to this module.
The potential benefit to have separate modules `openj9.management` and  `jdk.management` is that an application requiring `jdk.management` might choose to not require `openj9.management` hence not taxed the footprint of `openj9.management`.
If we don't expect such use scenario, i.e., an application requiring `jdk.management` always requires `openj9.management` as well, these two modules should be merged indeed.

File: jcl/.classpath [hunk 1]
```diff
@@ -1,3 +1,27 @@
...
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+ -->
+
```
Comment (keithc-ca) [inline jcl/.classpath]: The copyright notice cannot precede the `<?xml?>` line below.

Comment (mikezhang1234567890) [reply jcl/.classpath]: Thanks, fixed.

**Comment by DanHeidinga** [comment]
> If we don't expect such use scenario, i.e., an application requiring jdk.management always requires openj9.management as well, these two modules should be merged indeed.

It's a compatibility issue - we want applications that have opted into jdk.management to get OpenJ9's jdk-specfic mxbeans as well rather than having to conditionally add an extra module when running on OpenJ9.

File: jcl/src/openj9.sfj/share/classes/module-info.java [hunk 2]
```diff
@@ -1,6 +1,6 @@
 /*[INCLUDE-IF Sidecar19-SE]*/
 /*******************************************************************************
- * Copyright (c) 2016, 2017 IBM Corp. and others
+ * Copyright (c) 2017, 2017 IBM Corp. and others
```
Comment (AdamBrousseau) [inline jcl/src/openj9.sfj/share/classes/module-info.java]: This looks incorrect

Comment (mikezhang1234567890) [reply jcl/src/openj9.sfj/share/classes/module-info.java]: This one isn't actually a rename like Github thinks. I added the `module-info.java` file for this module to OpenJ9 so our old build system can process it and remove the line that exports to the old `com.ibm.management` name. So this file is completely new to OpenJ9.

**Comment by JasonFengJ9** [comment]
@mikezhang1234567890 pls advise if the PR is ready for reviewing.

**Comment by mikezhang1234567890** [comment]
@JasonFengJ9 PR is ready for reviewing now.

File: jcl/j9modules.txt [hunk 3]
```diff
@@ -44,8 +44,9 @@ java.management
 # requires java.management
 jdk.management
 
-# requires jdk.management
-com.ibm.management
+#requires jdk.management, processed to require the right module name
```
Comment (JasonFengJ9) [inline jcl/j9modules.txt]: "processed to require the right module name" doesn't seem provide more info.

File: jcl/j9modules.xml [hunk 4]
```diff
@@ -38,6 +38,12 @@
 			<case value="java.management">
 				<removeexport package="com.ibm.lang.management" />
 				<removeexport package="com.ibm.virtualization.management" />
+				<removeexport package="com.ibm.java.lang.management.internal" />
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: Can `exports com.ibm.java.lang.management.internal to jdk.management;` be removed as well?

File: jcl/j9modules.xml [hunk 5]
```diff
@@ -38,6 +38,12 @@
...
+				<removeopens package="java.lang.management" />
+
+				<replaceregexp file="${module.dir}/classes/module-info.java" replace="" flags="gs">
+					<regexp pattern="com\.ibm\.management," />
+				</replaceregexp>
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: Why is this needed?

File: jcl/j9modules.xml [hunk 6]
```diff
@@ -51,6 +57,8 @@
 				<removeexport package="com.ibm.dataaccess" />
 				<removeexport package="com.ibm.jvm" />
 				<removeexport package="com.ibm.oti.shared" />
+				<removeexport package="com.ibm.oti.util" />
+				<removeexport package="com.ibm.oti.vm" />
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: Again, these two packages are exported to modules other than `com.ibm.management`.

File: jcl/j9modules.xml [hunk 7]
```diff
@@ -83,6 +91,19 @@
 					</else>
 				</if>
 			</case>
+			<case value="jdk.jconsole">
+				<replaceregexp file="${module.dir}/classes/module-info.java" replace="jdk.management" flags="gs">
+					<regexp pattern="com\.ibm\.management" />
+				</replaceregexp>
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: Can we avoid this kind of universal replacement task and use some `macrodef` like `replaceRequires`.

File: jcl/j9modules.xml [hunk 8]
```diff
@@ -83,6 +91,19 @@
...
+			</case>
+			<case value="jdk.management">
+				<removeexport package="com.sun.management.internal" />
+			</case>
+			<case value="openj9.sfj">
+				<replaceregexp file="${module.dir}/classes/module-info.java" replace="jdk.management" flags="gs">
+					<regexp pattern="com\.ibm\.management" />
+				</replaceregexp>
+			</case>
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: Same comment here like above.

File: jcl/j9modules.xml [hunk 9]
```diff
@@ -104,6 +125,22 @@
...
+				<string value="@{package}" />
+				<filterchain>
+					<replaceregex pattern="\." replace="\\\\." flags="g" />
+				</filterchain>
+			</loadresource>
+			<replaceregexp file="${module.dir}/classes/module-info.java" replace="" flags="gs">
+				<regexp pattern="\s*\bopens\s+${regexp.@{package}}\b[^;]*;" />
+			</replaceregexp>
+		</sequential>
+	</macrodef>
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: Can `removeexport` and `removeopens` be generalized to a single `macrodef`?

File: jcl/src/java.management/share/classes/module-info.java.extra [hunk 10]
```diff
@@ -22,10 +22,9 @@
  *******************************************************************************/
 module java.management {
 	requires java.logging;
-	exports com.ibm.java.lang.management.internal to com.ibm.management, jdk.management;
-	exports sun.management.spi to com.ibm.management;
```
Comment (JasonFengJ9) [inline jcl/src/java.management/share/classes/module-info.java.extra]: is `sun.management.spi` to be exported to `jdk.management` now?

File: jcl/src/jdk.jconsole/share/classes/module-info.java.extra [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+module jdk.jconsole {
+}
```
Comment (JasonFengJ9) [inline jcl/src/jdk.jconsole/share/classes/module-info.java.extra]: why we need an empty `module-info.java` for `jdk.jconsole`?

Comment (mikezhang1234567890) [reply jcl/src/jdk.jconsole/share/classes/module-info.java.extra]: For `jdk.console` and openj9.sfj`, when we compose, they mention `com.ibm.management` by name. Adding the `module-info.java.extra` file is necessary for these modules to be processed by j9modules.xml

File: jcl/j9modules.txt (hunk 3 above)
Comment (mikezhang1234567890) [reply jcl/j9modules.txt]: Would "processed by j9modules.xml file to require the right module name" be better?

File: jcl/src/openj9.sfj/share/classes/module-info.java [hunk 12]
```diff
@@ -20,15 +20,5 @@
...
- * OpenJ9 management extensions.
- */
-module com.ibm.management {
-	requires transitive java.management;
-	requires transitive jdk.management;
-	requires java.logging;
-	exports com.ibm.lang.management;
-	exports com.ibm.virtualization.management;
-	provides sun.management.spi.PlatformMBeanProvider with com.ibm.lang.management.internal.PlatformMBeanProvider;
+module openj9.sfj {
```
Comment (JasonFengJ9) [inline jcl/src/openj9.sfj/share/classes/module-info.java]: Again, what's the usage of this empty `module-info.java`?

File: jcl/src/java.management/share/classes/module-info.java.extra (hunk 10 above)
Comment (mikezhang1234567890) [reply jcl/src/java.management/share/classes/module-info.java.extra]: It is already exported to `jdk.management` by the original `java.management` module.

File: jcl/j9modules.xml (hunk 9 above)
Comment (mikezhang1234567890) [reply jcl/j9modules.xml]: They can't be combined since there are modules that both export and open a package. In the case of this PR, `java.management` exports `java.lang.management` and opens it to `jdk.management`.

File: jcl/j9modules.xml (hunk 4 above)
Comment (mikezhang1234567890) [reply jcl/j9modules.xml]: This is to remove the export statement containing `com.ibm.management` during composing. The export statement without the old package name is added back in by `module-info.java.extra`.

File: jcl/j9modules.xml (hunk 6 above)
Comment (mikezhang1234567890) [reply jcl/j9modules.xml]: This is to remove the export statements containing `com.ibm.management` during composing. The export statements without the old package name are added back in by `module-info.java.extra` files.

**Comment by mikezhang1234567890** [comment]
@JasonFengJ9 can I get a re-review?

File: jcl/j9modules.xml [hunk 13]
```diff
@@ -83,28 +89,69 @@
...
 				</filterchain>
 			</loadresource>
 			<replaceregexp file="${module.dir}/classes/module-info.java" replace="" flags="gs">
-				<regexp pattern="\s*\bexports\s+${regexp.@{package}}\b[^;]*;" />
+				<regexp pattern="\s*\b@{action}\s+${regexp.@{package}}\b[^;]*;" />
 			</replaceregexp>
 		</sequential>
 	</macrodef>
 
+	<macrodef name="replaceexportsoropens">
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: When the `module.old` is the only module in `exports|opens` clauses, this `replaceexportsoropens` effectively becomes `removeexportsoropens`, can we combine these two `macrodef` together as well?

File: jcl/j9modules.xml [hunk 14]
```diff
@@ -83,28 +89,69 @@
...
 		<sequential>
-			<echo message="Removing export of package @{package}" />
+			<echo message="Removing @{action} of package @{package}" />
 			<loadresource property="regexp.@{package}">
 				<string value="@{package}" />
 				<filterchain>
 					<replaceregex pattern="\." replace="\\\\." flags="g" />
 				</filterchain>
 			</loadresource>
 			<replaceregexp file="${module.dir}/classes/module-info.java" replace="" flags="gs">
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: `${module.dir}/classes/module-info.java` might be defined as a variable and used across this xml file.

File: runtime/jvm.xml.ftl [hunk 15]
```diff
@@ -1282,7 +1282,15 @@
...
 						</if>
-
+		
+						<if>
+							<not>
+								<available file="${r"${j9jcl.module.info}/module-info.java"}" type="file" />
+							</not>
+							<then>
+								<copy file="${r"${output}/${module.name}_extracted/classes/module-info.java"}" tofile="${r"${j9jcl.module.info}/module-info.java"}" /> 
+							</then>
```
Comment (JasonFengJ9) [inline runtime/jvm.xml.ftl]: A `else` block can be followed to include following target `-merge-module-info`, otherwise it just merges two identical files unnecessarily.

File: jcl/j9modules.xml (hunk 13 above)
Comment (mikezhang1234567890) [reply jcl/j9modules.xml]: `removeexportsoropens` removes lines based on the package name being exported while `replaceexportsoropens` uses module names. So `replaceexportsoropens` doesn't remove exports or opens lines that don't have a target module or multiple target modules. 

Should I rename the macro `replaceexportsoropensmodules` for clarity?

**Comment by AdamBrousseau** [comment]
Jenkins copyright check
Jenkins line endings check

**Comment by AdamBrousseau** [comment]
Jenkins copyright check

**Comment by AdamBrousseau** [comment]
Jenkins line endings check

**Comment by mikezhang1234567890** [comment]
Updated commit message to mention the macros added to `j9modules.xml`.
@JasonFengJ9 can I get a re-review?

File: runtime/jvm.xml.ftl (hunk 15 above)
Comment (JasonFengJ9) [reply runtime/jvm.xml.ftl]: Pls add comments that the module-info might have been modified such that duplication cleaning up is necessary, also list the target modules by this change.

File: jcl/j9modules.xml [hunk 16]
```diff
@@ -115,10 +167,28 @@
...
 				<regexp pattern="\s*\bprovides\s+${regexp.@{interface}}\s+with[^;]*;" />
 			</replaceregexp>
 		</sequential>
 	</macrodef>
+	
+	<macrodef name="replacerequires">
+		<attribute name="module.old" />
+		<attribute name="module.new" />
+		<sequential>
+			<echo message="Replaces requiring s@{module.old} with @{module.new}" />
```
Comment (JasonFengJ9) [inline jcl/j9modules.xml]: An extra `s` here, a typo?

**Comment by JasonFengJ9** [comment]
lgtm

**Comment by mikezhang1234567890** [comment]
Fixed an issue with ant properties persisting through macro calls with the newly added macros. Tested running an internal build, which went through fine.

@JasonFengJ9 @pshipton can you review again?

**Comment by JasonFengJ9** [comment]
@mikezhang1234567890 pls rebase this PR with latest master branch for easier comparison.

**Comment by JasonFengJ9** [comment]
> _Fixed an issue with ant properties persisting through macro calls with the newly added macros._ 

This change has been released via https://github.com/eclipse/openj9/pull/661
LGTM

File: jcl/src/java.management/share/classes/module-info.java.extra [hunk 17]
```diff
@@ -22,10 +22,9 @@
...
+	exports com.ibm.java.lang.management.internal to jdk.management;
 	/*[IF Open-Module-Support]*/
-	opens java.lang.management to com.ibm.management;
+	opens java.lang.management to jdk.management;
```
Comment (pshipton) [inline jcl/src/java.management/share/classes/module-info.java.extra]: This isn't something OpenJ9 should be adding. Please track down what needs this so we can determine what needs to be done to avoid it.

Comment (mikezhang1234567890) [reply jcl/src/java.management/share/classes/module-info.java.extra]: Reflection is used in `ExtendedThreadInfoImpl.java` to access the `nativeTId` field in ThreadInfo object from the `java.management` module for use.

Comment (pshipton) [reply jcl/src/java.management/share/classes/module-info.java.extra]: We can create an internal access class to query this, and avoid the use of reflection and having to open the module.

Comment (pshipton) [reply jcl/src/java.management/share/classes/module-info.java.extra]: See java.lang.VMAccess for an example

File: jcl/src/java.management/share/classes/java/lang/management/ManagementAccessControl.java [hunk 18]
```diff
@@ -0,0 +1,53 @@
...
+package java.lang.management;
+
+import com.ibm.oti.util.Msg;
+
+/**
+ * This class is used to give access to APIs which give access to
+ * package private methods in the java.management module similar to
+ * sun.misc.SharedSecrets
+ */
+public class ManagementAccessControl {
```
Comment (pshipton) [inline jcl/src/java.management/share/classes/java/lang/management/ManagementAccessControl.java]: Additional public classes which aren't part of the java specification cannot be added to the java.lang.management package, this is a violation of the specification. Put these public classes in the com.ibm.lang.management.internal package.

File: jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementAccessControl.java [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+	
+	
+	private static ThreadInfoAccess threadInfoAccess;
+	
+	/**
+	 * Sets the access object for ThreadInfo. This should only be called once.
+	 */
+	public static void setThreadInfoAccess(ThreadInfoAccess access) {
+		if (threadInfoAccess != null) {
+			throw new SecurityException(Msg.getString("K05ba")); //$NON-NLS-1$
```
Comment (JasonFengJ9) [inline jcl/src/java.management/share/classes/com/ibm/java/lang/management/internal/ManagementAccessControl.java]: Need add following before the `throw`
````
/*[MSG "K05ba", "Cannot set access twice"]*/
````

File: jcl/src/java.management/share/classes/java/lang/management/ThreadInfo.java [hunk 20]
```diff
@@ -39,14 +40,18 @@
...
+	static {
+		ManagementAccessControl.setThreadInfoAccess(new ThreadInfoAccessImpl());
+	}
 
 	private long threadId;
 
 	/**
-	 * @brief Do /not/ rename this field; it is initialized but not used within the 
-	 * class's code (hence annotated "unused"), whereas, it is accessed via Reflection.
+	 * @brief Do /not/ rename this field; it is accessed via a package private
```
Comment (pshipton) [inline jcl/src/java.management/share/classes/java/lang/management/ThreadInfo.java]: The `Do /not/ rename this field` part can be removed, as its now accesses via getNativeThreadId() instead of reflection.

Comment (pshipton) [reply jcl/src/java.management/share/classes/java/lang/management/ThreadInfo.java]: You could just remove the entire comment, as getNativeThreadId() is commented.
